### PR TITLE
[ISSUE #105] Add PullConsumer. resolve #105

### LINKED ISSUE #105: Add PullConsumer
**FEATURE REQUEST**

1. Please describe the feature you are requesting.
Add PullConsumer

2. Provide any additional detail on your proposed use case for this feature.
Add PullConsumer

2. Indicate the importance of this issue to you (blocker, must-have, should-have, nice-to-have). Are you currently using any workarounds to address this issue?
must-have 

4. If there are some sub-tasks using -[] for each subtask and create a corresponding issue to map to the sub task:

### PR DESCRIPTION
**Pull request by xujianhai666**
## What is the purpose of the change

Add PullConsumer

## Brief changelog

Add PullConsumer

## Verifying this change

add example

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -54,7 +54,11 @@ type PullConsumer interface {
 	Shutdown() error
 	Pull(context.Context, string, consumer.MessageSelector, int) (*primitive.PullResult, error)
 	PullFrom(context.Context, *primitive.MessageQueue, int64, int) (*primitive.PullResult, error)
-	// only update in memory
-	UpdateOffset(primitive.MessageQueue, int64) error
+	CurrentOffset(queue *primitive.MessageQueue) (int64, error)
```
Comment (wenfengwang) [inline api.go]: remove ‘queue’

Comment (xujianhai666) [reply api.go]: @wenfengwang done

File: consumer/pull_consumer.go [hunk 2]
```diff
@@ -19,40 +19,104 @@ package consumer
...
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrMQEmpty = errors.New("MessageQueue is nil")
+	ErrOffset  = errors.New("offset < 0")
+	ErrNumbers = errors.New("numbers < 0")
 )
 
 type PullConsumer interface {
```
Comment (wenfengwang) [inline consumer/pull_consumer.go]: Don't re-define PullConsumer Interface, please refer `api.go`

File: consumer/consumer.go [hunk 3]
```diff
@@ -247,18 +253,65 @@ type defaultConsumer struct {
...
 }
 
-func (dc *defaultConsumer) persistConsumerOffset() {
+func (dc *defaultConsumer) start() error {
+
+	if dc.model == Clustering {
+		// set retry topic
+		retryTopic := internal.GetRetryTopic(dc.consumerGroup)
+		dc.subscriptionDataTable.Store(retryTopic, buildSubscriptionData(retryTopic,
+			MessageSelector{TAG, _SubAll}))
```
Comment (wenfengwang) [inline consumer/consumer.go]: Don't use anonymous initialization

File: consumer/pull_consumer.go [hunk 4]
```diff
@@ -19,40 +19,104 @@ package consumer
...
+
+	srvs, err := internal.NewNamesrv(defaultOpts.NameServerAddrs...)
+	if err != nil {
+		return nil, errors.Wrap(err, "new Namesrv failed.")
+	}
+	internal.RegisterNamsrv(srvs)
+
+	dc := &defaultConsumer{
+		consumerGroup: defaultOpts.GroupName,
+		cType:         _PushConsume,
```
Comment (wenfengwang) [inline consumer/pull_consumer.go]: _PullConsumer

File: consumer/push_consumer.go [hunk 5]
```diff
@@ -126,20 +126,18 @@ func (pc *pushConsumer) Start() error {
...
 
-		pc.client = internal.GetOrNewRocketMQClient(pc.option.ClientOptions)
-		if pc.model == Clustering {
-			pc.option.ChangeInstanceNameToPID()
-			pc.storage = NewRemoteOffsetStore(pc.consumerGroup, pc.client)
-		} else {
-			pc.storage = NewLocalFileOffsetStore(pc.consumerGroup, pc.client.ClientID())
+		err := pc.client.RegisterConsumer(pc.consumerGroup, pc)
+		if err != nil {
+			pc.state = internal.StateCreateJust
```
Comment (wenfengwang) [inline consumer/push_consumer.go]: Add new state `StateStartFailed`

File: examples/consumer/pull/main.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/apache/rocketmq-client-go/rlog"
+	"github.com/apache/rocketmq-client-go/utils"
+)
+
+func main() {
+	c, err := consumer.NewPullConsumer(consumer.WithGroupName("testGroup"), consumer.WithNameServer([]string{"127.0.0.1:9876"}))
+	if err != nil{
+		rlog.Fatal("fail to new pullConsumer: %v", err)
+	}
+	c.Start()
```
Comment (wenfengwang) [inline examples/consumer/pull/main.go]: error handler

File: consumer/pull_consumer.go (hunk 2 above)
Comment (xujianhai666) [reply consumer/pull_consumer.go]: @wenfengwang if do like that, we will introduce the problem "import cycle", I put that in utils/errors.go.

File: consumer/consumer.go (hunk 3 above)
Comment (xujianhai666) [reply consumer/consumer.go]: done

File: consumer/push_consumer.go (hunk 5 above)
Comment (xujianhai666) [reply consumer/push_consumer.go]: done

File: examples/consumer/pull/main.go (hunk 6 above)
Comment (xujianhai666) [reply examples/consumer/pull/main.go]: it is the root call, so if err, we just panic, rlog.Fatal cloud print error and pnic.

File: consumer/pull_consumer.go (hunk 4 above)
Comment (xujianhai666) [reply consumer/pull_consumer.go]: done

**Comment by wenfengwang** [comment]
LGTM, @ShannonDing @zongtanghu @jonnxu please merge the PR

**Pull request merged**